File: 2024/01/01.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
	"strings"
)

func parseIntField(in string) int {
	num, err := strconv.Atoi(in)
	if err != nil {
		panic(err)
	}
	return num
}

func parseLists() ([]int, []int) {
	var a, b []int
	file, err := os.Open("input.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		parts := strings.Fields(scanner.Text())
		if len(parts) == 2 {
			a = append(a, parseIntField(parts[0]))
			b = append(b, parseIntField(parts[1]))
		}
	}
	return a, b
}

func abs(i int) int {
	if i < 0 {
		return -i
	}
	return i
}

func measureDistance(a, b []int) {
	distance := 0
	sort.Slice(a, func(i, j int) bool {
		return a[i] < a[j]
	})
	sort.Slice(b, func(i, j int) bool {
		return b[i] < b[j]
	})
	for i := range a {
		distance += abs(a[i] - b[i])
	}

	fmt.Printf("distance: %d\n", distance)
}

func measureSimilarityScore(a, b []int) {
	score := 0
	for _, k := range a {
		for _, j := range b {
			if k == j {
				score += k
			}
		}
	}
	fmt.Printf("similarity score: %d\n", score)
}

func main() {
	first, second := parseLists()
	measureDistance(first, second)
	measureSimilarityScore(first, second)
}

File: 2024/01/01.py
def absolute_value(x):
    return x if x >= 0 else -x

def measure_distance(first, second):
    distance = 0
    first.sort()
    second.sort()
    for i in range(0, len(first), 1):
        distance += absolute_value(first[i] - second[i])
    print(f"distance: {distance}")

def measure_similarity(first, second):
    similarity_score = 0
    for i in first:
        similarity_score += i * second.count(i)
    print(f"similarity score: {similarity_score}")

def get_lists():
    first, second = [], []
    with open("input.txt", "r") as f:
        for line in f:
            parts = line.split()
            if len(parts) == 2:
                first.append(int(parts[0]))
                second.append(int(parts[1]))
    return first, second

if __name__ == "__main__":
    first, second = get_lists()
    measure_distance(first, second)
    measure_similarity(first, second)